File: NUBAN.py

def __main__():
    '''Program to validate NUBAN account'''
    print("Hello!")
    code = input("Enter the bank code:_")
    account_number = input("Enter the account number:_")

    if validate_nuban_account(code, account_number):
        response = "Valid"
    else:
        response = "Invalid"

    print("The account number is {}.".format(response))

def validate_nuban_account(bank_code, account_number):
    '''Validate Account Function'''
    result = False

    if len(bank_code.rstrip()) == 3 and len(account_number.rstrip()) == 10:
        nuban = bank_code + account_number[:-1]

        check_digit = int(account_number[9:10])
        nuban_array = []

        for i in range(0, len(nuban)):
            nuban_array.append(int(str(nuban[i])))
        nuban_sum = (nuban_array[0]*3 + nuban_array[1]*7 + nuban_array[2]*3
                     + nuban_array[3]*3 + nuban_array[4]*7 + nuban_array[5]*3
                     + nuban_array[6]*3 + nuban_array[7]*7 + nuban_array[8]*3
                     + nuban_array[9]*3 + nuban_array[10]*7 + nuban_array[11]*3)

        cal_check_digit = 10 - (nuban_sum%10)
        if cal_check_digit == 10:
            cal_check_digit = 0

        if check_digit == cal_check_digit:
            result = True
    return result

if __name__ == '__main__':
    __main__()

File: NUBAN.go
package main

import (
	"bufio"
    "fmt"
    "os"
    "strings"
    "strconv"
)

func main() {
	fmt.Println("Hello there")
	fmt.Println("Enter the bank sort code")
	reader := bufio.NewReader(os.Stdin)
	sortcode, _ := reader.ReadString('\n')
	fmt.Println(sortcode)
	fmt.Println("Enter the account number")
	account_number, _ := reader.ReadString('\n')
	fmt.Println(account_number)
	response := ""
	response_bool := validate_nuban_account(sortcode, account_number)
	if(response_bool == true){
		response = "The account number is Valid"
	}else{
		response = "The account number is Invalid"
	}

    println(response)
}

func validate_nuban_account(bankCode string, accountNumber string) bool {
       result := false

        if (len(strings.TrimSpace(bankCode)) == 3 && len(strings.TrimSpace(accountNumber)) == 10){
            nuban := bankCode + accountNumber[:8]
            checkDigit, err := strconv.Atoi(accountNumber[:8])
		    if err != nil {
		        // account number cannot be parsed to integer
		        fmt.Println(err)
		        os.Exit(2)
		    }
		    nubanArray := make([]int, len(nuban))
		    for i := 0; i < len(nuban); i++ {
		    	str := convert(nuban[i])
		    	nubanArray[i], err = strconv.Atoi(str)
		    }
            fmt.Println(len(nubanArray))
            nubanSum := (nubanArray[0] * 3) + (nubanArray[1] * 7) + (nubanArray[2] * 3) + (nubanArray[3] * 3) +
                    (nubanArray[4] * 7) + (nubanArray[5] * 3) + (nubanArray[6] * 3) + (nubanArray[7] * 7) +
                    (nubanArray[8] * 3) + (nubanArray[9] * 3) + (nubanArray[10] * 7) + (nubanArray[11] * 3)
            calCheckDigit := 10 - (nubanSum % 10)

            if(checkDigit != 10){
            	 calCheckDigit = calCheckDigit
            }else{
            	calCheckDigit = 0
            }
            result = (checkDigit == calCheckDigit)
        }

   	return result

}

func convert( b byte ) string {   
    s := strconv.Itoa(int(b))
    return s
}